### PR TITLE
Add pre-commit to generate ARM template

### LINKED ISSUE #519: Generate the json deployment file in the pre-commit hook
Utilise the pre-commit hook and bicep to generate the json file so that they are not out of sync with each other.

### PR DESCRIPTION
**Pull request by ross-p-smith**
# Purpose
Fixes #519

- Add a script that will call `az bicep` to format and create a json representation of the bicep. This json will be used by the `Deploy to Azure` button and for those that are ignoring the existence of the devcontainer and `az up`.
- Then call this script as part of the pre-commit hook so that it is called each time.

Here is a screen shot where I only changed the bicep - but when I pushed, the changes were also reflected in the json too.
image

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -10,3 +10,15 @@ repos:
...
+    hooks:
+      - id: bicep
+        name: bicep
+        description: Lint and build Bicep files
+        entry: ./infra/generate_arm_template.sh
+        language: script
+        files: \.bicep$
+        require_serial: true
+        args: # Bicep files that we want to generate ARM templates from
+          - -f=./infra/main.bicep
```
Comment (martinpeck) [inline .pre-commit-config.yaml:24]: Can you add a final new-line please?
Pretty please?
🥺

File: README.md [hunk 2]
```diff
@@ -163,7 +163,7 @@ az deployment sub create --subscription <subscription-id> --location <location>
 
 Or
 
-To deploy with isolated network use following command. (Replace the parameter values where needed). If you do not fill in all three, the deployment will not deploy this to an isolated network.
+To deploy with isolated network use following command. Replace the parameter values with the specifics of your isolated network. You **must** supply all three parameters (i.e. `vnetAddressSpace`, `proxySubnetAddressSpace` and `subnetAddressSpace`) if you wish to deploy to an isolated network.
```
Comment (martinpeck) [inline README.md:166]: lovely....couldn't have reworded it better myself!

File: infra/generate_arm_template.sh [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+wait
+
+git add ${FILES[@]}
+
+# Build the templates
+for TEMPLATE in ${TEMPLATES[@]}; do
+    az bicep build -f $TEMPLATE
+    git add "${TEMPLATE%.bicep}.json" # Change the extension from .bicep to .json
+done
```
Comment (martinpeck) [inline infra/generate_arm_template.sh:29]: Give it another smash on the old RETURN key and add a lovely final new-line, please.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (martinpeck) [inline .pre-commit-config.yaml:24]: Will this cause the main.bicep file to be converted to ARM even if it wasn't changed?
e.g. if I modify `keyvault.bicep`

Is that what we want? I assume so, but want to check

Comment (ross-p-smith) [reply .pre-commit-config.yaml]: Here - I have changed a word in a child template and it has still regenerated the json
image

**Review by martinpeck — APPROVED** [review]

**Pull request merged**